### ISSUE TITLE
Deprecate `hash_map` and `hash_set` strategies?

### ISSUE THREAD

**Issue by thomaseizinger** [proptest-rs/proptest#494]
I've noticed that the `hash_map` and `hash_set` strategies can be quite severe footguns. Iteration order of `HashMap`s and `HashSet`s is non-deterministic. Hence, composing more strategies on top of a `hash_map` with `.prop_flat_map` will most likely yield non-deterministic behaviour and completely break the idea of reproducing a failing test case based on its seed.

I've spent many hours debugging my tests (`proptest-state-machine` specifically) until I finally got to the bottom of this.

I am not sure if deprecation is the right call but I think it should be called out or warned about somewhere.

**Comment by giluis** [comment]
What should be used instead of these builtin strategies? I'm using them at work

**Comment by thomaseizinger** [comment]
> What should be used instead of these builtin strategies? I'm using them at work

`BTreeMap`s and `BTreeSet`s are alternatives that offer deterministic ordering.

**Comment by skaunov** [comment]
I came here for quite same reason, though was lucky to notice that those `Iterator` behaves arbitrary on each instantiation, hence undermining reproducibility. I'd say deprecation of the default strategies won't help at all as this issue affects any `Hash` based collection which is instantiated in a `proptest`. =( Is it possible to provide a drop-in replacement type with the crate and note that in the docs? I guess it would leverage <https://doc.rust-lang.org/std/collections/struct.HashSet.html#usage-in-const-and-static>.

Actually I think it would be nice if a macro would raise a warning on such a collection usage...

**Comment by thomaseizinger** [comment]
> I'd say deprecation of the default strategies won't help at all as this issue affects any `Hash` based collection which is instantiated in a `proptest`. =(

I think simple usage of these types within your code shouldn't be quite as bad as using the strategies. Yes your code won't be 100% deterministic in how it executes but at least it will be initialised with the same values.

Using these types in strategies is much worse because you won't even get the same parameters for two runs of the same seed.

**Comment by skaunov** [comment]
I'm afraid shrinking is "out of the window" in this case.

But I formulated my comment awfully: deprecating `Hash` based strategies in spite of `BTree` based is a great step forward -- I agree on this actually.

Also the problem I was thinking on is getting deeper when _something_ is using `Hash` underneath. That could be a case, and deserves at least a mention in the book, and a macro warning for cases where it is in clear worth that too I guess.

**Comment by thomaseizinger** [comment]
I ended up adding tests that ensure my strategies are deterministic based on a given seed: https://github.com/firezone/firezone/blob/2f43bb9110796da2dfcd0f5913ff53c8d3922e11/rust/connlib/tunnel/src/tests.rs#L129-L148

Something like that could be added as a recommendation in the book or perhaps even auto-generated as part of the macro?

**Comment by skaunov** [comment]
> auto-generated as part of the macro?

yeah, better like an additional tool --- I can imagine lots of people who won't be willing to pay for this each test run, esp in continuous settings
